src/Bases/ModelBase.py

```diff
@@ -0,0 +1,166 @@
+      @ Out, None
+    """
+    needDict = self.getInitParams()
+    needDict = self.loadVariables(needDict, inputDict)
+    self.setParams(needDict)
+    self._checkInputParams(needDict)
+
+  def _checkInputParams(self, needDict):
+    """
+      Method to check input parameters
+      @ In, needDict, dict, dictionary of required parameters
+      @ Out, None
+    """
+    pass
+
+  def isDynamic(self):
+    """
+      This method is utility function that tells if the reliability model is able to
+      treat dynamic data on its own or not
+      @ In, None
+      @ Out, isDynamic, bool, True if the reliability model is able to treat dynamic data, False otherwise
+    """
+    return self._dynamicHandling
+
+  def setVariable(self, value):
+    """
+      Set value if a float/int/list is provided in the node text, othersise treat the provided value as RAVEN variable
+      @ In, value, str or float or list, the value of given variable
+      @ Out, ret, str or float or numpy.array, the recasted value
+    """
+    ret = None
+    # multi-entry or single-entry?
+    if len(value) == 1:
+      if not utils.isAFloatOrInt(value[0]):
+        ret = value[0]
+      else:
+        ret = np.atleast_1d(value)
+    else:
+      # should be floats; InputData assures the entries are the same type already
+      if not utils.isAFloatOrInt(value[0]):
+        raise IOError('Multiple non-number entries are found, but require either a single variable name or multiple float entries: {}'.format(value))
+      ret = np.asarray(value)
+    return ret
+
+  def loadVariables(self, need, inputDict):
+    """
+      Load the values of variables that is generated by RAVEN
+      @ In, need, dict, the dict of parameters
+      @ In, inputDict, dict, the dict of parameters that is provided from other sources
+      @ Out, need, dict, the dict of parameters updated with variables
+    """
+    # load variable values from variables as needed
+    for key, val in need.items():
+      if utils.isAString(val):
+        value = inputDict.get(val, None)
+        if value is None:
+          raise KeyError('Looking for variable "{}" to fill "{}" but not found among variables!'.format(val, key))
+        need[key] = np.atleast_1d(value)
+    return need
+
+  def getInitParams(self):
+    """
+      Get the initial parameters
+      @ In, None
+      @ Out, self._variableList, dict, list of variables
+    """
+    ## copy is required, since we are using the initial setup of self._variableDict
+    ## to tell the system which variables to update on the fly.
+    return copy.copy(self._variableDict)
+
+  def setParams(self, paramDict):
+    """
+      Set the parameters from a given dictionary.
+      @ In, paramDict, dict, settings
+      @ Out, None
+    """
+    for key, val in paramDict.items():
+      if key in self.__dict__.keys():
+        setattr(self, key, val)
+      else:
+        print('WARNING: Variable "{}" is not defined in class "{}"!'.format(key, self.name))
```

Comment:
should we raise a KeyError?

Reply:
I change it to KeyError now. For our current applications, it is more safe to do so. In the future, if our models also want to acquire the meta data information, and this data was not a self variable in the model, we can remove the check. I put the check there is mainly for the developer to remind them if they forget to add the on-the-fly updated variable to self._variableDict.